solver/src/main/java/org/chocosolver/solver/constraints/binary/PropSquare.java

```diff
@@ -117,57 +108,95 @@ protected void updateLowerBoundofX() throws ContradictionException {
             for (int value = vars[0].getLB(); value <= ub; value = vars[0].nextValue(value)) {
-                if (!(MathUtils.isPerfectSquare(value) &&
-                        (vars[1].contains(floor_sqrt(value)) || vars[1].contains(-floor_sqrt(value))))) {
+                int sqrt = floor_sqrt(value);
+                if (!vars[1].contains(sqrt) && !vars[1].contains(-sqrt)) {
                     vrms.add(value);
                 }
             }
-            return vars[0].removeValues(vrms, this);
-        } else if (vars[0].hasEnumeratedDomain()) {
+            impact |= vars[0].removeValues(vrms, this);
+        } else {
             int value = vars[0].getLB();
             int nlb = value - 1;
             while (nlb == value - 1) {
-                if (!vars[1].contains(floor_sqrt(value)) && !vars[1].contains(-floor_sqrt(value))) {
+                int sqrt = floor_sqrt(value);
+                if (!vars[1].contains(sqrt) && !vars[1].contains(-sqrt)) {
                     nlb = value;
                 }
                 value = vars[0].nextValue(value);
             }
-            boolean filter = vars[0].updateLowerBound(nlb, this);
+            impact |= vars[0].updateLowerBound(nlb, this);
 
             value = vars[0].getUB();
             int nub = value + 1;
             while (nub == value + 1) {
-                if (!vars[1].contains(floor_sqrt(value)) && !vars[1].contains(-floor_sqrt(value))) {
+                int sqrt = floor_sqrt(value);
+                if (!vars[1].contains(sqrt) && !vars[1].contains(-sqrt)) {
                     nub = value;
                 }
                 value = vars[0].previousValue(value);
             }
-            return filter | vars[0].updateUpperBound(nub, this);
+            impact |= vars[0].updateUpperBound(nub, this);
         }
-        return false;
+        return impact;
     }
 
     protected boolean updateLowerBoundofY() throws ContradictionException {
-        return vars[1].updateLowerBound(-ceil_sqrt(vars[0].getUB()), this);
+        if (vars[1].getLB() >= 0) {
+            return vars[1].updateLowerBound(ceil_sqrt(vars[0].getLB()), this);
+        } else {
+            return vars[1].updateLowerBound(-floor_sqrt(vars[0].getUB()), this);
+        }
     }
 
     protected boolean updateUpperBoundofY() throws ContradictionException {
-        return vars[1].updateUpperBound(floor_sqrt(vars[0].getUB()), this);
+        if (vars[1].getUB() < 0) {
+            return vars[1].updateUpperBound(-ceil_sqrt(vars[0].getLB()), this);
+        } else {
+            return vars[1].updateUpperBound(floor_sqrt(vars[0].getUB()), this);
+        }
     }
 
     protected boolean updateHolesinY() throws ContradictionException {
-        // remove intervals to deal with consecutive value removal and upper bound modification
-        if (bothEnum) {
+        if (vars[0].isInstantiatedTo(0)) {
+            return vars[1].instantiateTo(0, this);
+        }
```

Comment:
Is this useful ? If vars[0] is instantiated to 0, vars[1] should already be valued to 0 with updateLowerBoundofY() and updateUpperBoundofY().

It might be more useful to have an if statement returning false if vars[1] has not an enumerated domain.

Reply:
I has been introduce in a commit to fix a bug so, since it has no impact on performance, I preferred to keep it.